libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+	                                   &resample_frames, &ts_offset,
+	                                   (const uint8_t *const *) audio_data->data,
+	                                   (uint32_t) audio_data->frames);
+	if (!success) {
+		goto unlock;
+	}
+
+	bytes = monitor->bytes_per_frame * resample_frames;
+
+	if (muted) {
+		memset(resample_data[0], 0, bytes);
+	} else {
+		if (!close_float(vol, 1.0f, EPSILON)) {
+			process_volume(monitor, vol, resample_data,
+			               resample_frames);
+		}
+	}
+
+	circlebuf_push_back(&monitor->new_data, resample_data[0], bytes);
+	monitor->packets++;
+	monitor->frames += resample_frames;
+
+unlock:
+	pthread_mutex_unlock(&monitor->playback_mutex);
+	do_stream_write(param);
+}
+
+static void
+pulseaudio_stream_write(pa_stream *p, size_t nbytes, void *userdata)
+{
+	UNUSED_PARAMETER(p);
+	PULSE_DATA(userdata);
+
+	pthread_mutex_lock(&data->playback_mutex);
+	data->bytesRemaining += nbytes;
+	pthread_mutex_unlock(&data->playback_mutex);
+
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_server_info(pa_context *c, const pa_server_info *i,
+                                   void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	UNUSED_PARAMETER(userdata);
+
+	blog(LOG_INFO, "Server name: '%s %s'",
+	     i->server_name, i->server_version);
+
+	pulseaudio_signal(0);
+}
+
+static void
+pulseaudio_source_info(pa_context *c, const pa_source_info *i, int eol,
+                       void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	PULSE_DATA(userdata);
+	// An error occured
+	if (eol < 0) {
+		data->format = PA_SAMPLE_INVALID;
+		goto skip;
+	}
+	// Terminating call for multi instance callbacks
+	if (eol > 0)
+		goto skip;
+
+	blog(LOG_INFO, "Audio format: %s, %"
+		PRIu32
+		" Hz"
+			", %"
+		PRIu8
+		" channels",
+	     pa_sample_format_to_string(i->sample_spec.format),
+	     i->sample_spec.rate,
+	     i->sample_spec.channels);
+
+	pa_sample_format_t format = i->sample_spec.format;
+	if (pulseaudio_to_obs_audio_format(format) == AUDIO_FORMAT_UNKNOWN) {
+		format = PA_SAMPLE_S16LE;
+
+		blog(LOG_INFO, "Sample format %s not supported by OBS,"
+			"using %s instead for recording",
+		     pa_sample_format_to_string(i->sample_spec.format),
+		     pa_sample_format_to_string(format));
+	}
+
+	uint8_t channels = i->sample_spec.channels;
+	if (pulseaudio_channels_to_obs_speakers(channels) == SPEAKERS_UNKNOWN) {
+		channels = 2;
+
+		blog(LOG_INFO, "%c channels not supported by OBS,"
+			"using %c instead for recording",
+		     i->sample_spec.channels,
+		     channels);
+	}
+
+	data->format = format;
+	data->samples_per_sec = i->sample_spec.rate;
+	data->channels = channels;
+skip:
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_stop_playback(struct audio_monitor *monitor)
+{
+	if (monitor->stream) {
+		pa_stream_disconnect(monitor->stream);
+		pa_stream_unref(monitor->stream);
+		monitor->stream = NULL;
+	}
+
+	blog(LOG_INFO, "Stopped Monitoring in '%s'", monitor->device);
+	blog(LOG_INFO, "Got %"
```

Comment:
Lines 262-266 (this `blog` call) are broken up in a manner that makes them difficult to read. See the following examples:

https://github.com/jp9000/obs-studio/blob/e006d961a479699921323dc1350939221dc0ed58/plugins/linux-pulseaudio/pulse-input.c#L195
https://github.com/jp9000/obs-studio/blob/a2953d1161a3ed64946d2cfbd98cafcc04d19be8/libobs/obs-output.c#L304